---
title: "Visualization for cell-type composition and single-cell spatial maps of SWOT"
output:
  rmdformats::html_clean:
    highlight: kate
---

```{r setup, include=FALSE}
## Global options
knitr::opts_chunk$set(cache = TRUE, 
                      comment="", 
                      comment = "#>",
                      out.width = "100%",
                      warning = FALSE)
```

# SWOT

## Introduction
SWOT is a spatially weighted optimal transport method that integrates single-cell RNA sequencing data with spatial transcriptomics data for the inference of cell-type composition and single-cell spatial maps.
In this example, we will visualize the cell-type composition results and the inferred single-cell spatial maps by SOWT in pancreatic ductal adenocarcinoma (PDAC) dataset.


```{r warning = FALSE}
library(ggplot2)
library(dplyr)
```

## Read data for visualization
The results obtained from running SWOT in Python are located in folder "SWOT_files/CellMapping/", including:   
&emsp;&emsp;"Celltype_composition.csv": cell-type composition matrix;
&emsp;&emsp;"Cell_maps_xy.csv": cell spatial coordinates matrix for inferring the single-cell spatial map;
&emsp;&emsp;"Cell_maps_exp.csv": single-cell gene expression matrix for inferring single-cell spatial map.


```{r}
file_path <- getwd()
data_path <- paste0(file_path, '/Data_PDAC/')
swot_path <- paste0(data_path, '/SWOT_files/')
# data_path <- paste0(file_path, '/Data/')
# swot_path <- paste0(file_path, '/SWOT_files/')

ST_xy <- read.csv(paste0(data_path, 'PDAC_st_xy.csv'), header = T, row.names = 1, check.names = F)

SWOT_CTprop <- read.csv(paste0(swot_path, 'CellMapping/Celltype_composition.csv'), header = T, row.names = 1, check.names = F)
SWOT_Cellmapping <- read.csv(paste0(swot_path, 'CellMapping/Cell_maps_xy.csv'), header = T, row.names = 1, check.names = F)
SWOT_Cellmapping_exp <- read.csv(paste0(swot_path, 'CellMapping/Cell_maps_exp.csv'), header = T, row.names = 1, check.names = F)

color_Celltype <- c("#c1f1fc", "#FFFFCC", "#ffc2e5", "#CCFF99", "#CCCCFF",
                    "#81ceff", "#ffc93c", "#FF00FF", "#19e3b1", "#b61aae",
                    "#0000FF", "#d2ea32", '#f5c7f7', "#a0ac48", "#590d82",
                    "#0C7EF3", "#f9a270", "#ff5454", "#007f4e", "#5707F8")

source("Visualization_functions.R")

```

## Visualization of cell-type composition results
The visualization of cell-type composition results includes the depiction of the spatial probability distribution of each cell type and the spatial distribution of individual cell types.

### ScatterPie of cell-type compositions
```{r fig.width=10, fig.height=8}
SWOT_CTprop <- SWOT_CTprop[order(names(SWOT_CTprop))]
Plot_Scatterpie(SWOT_CTprop, ST_xy, 
                pie_pointsize=0.8, color_ct=color_Celltype)
```

### Spatial distribution of one cell type
```{r fig.width=6, fig.height=5}
# Plot the spatial distribution of cell type 'Cancer clone A'
Plot_CTProp(SWOT_CTprop, ST_xy,
            ct = 'Cancer clone A', point_size=4.5)
```



## Visualization of single-cell spatial maps results
For the inferred single-cell spatial map of PDAC dataset, we plotted the spatial distribution of cell types and the spatial gene expression patterns for individual cell types.

### Cell-type spatial distribution of single-cell spatial map
```{r fig.width=10, fig.height=8}
Celltypes <- colnames(SWOT_CTprop)
SWOT_Cellmapping$cs_type <- factor(SWOT_Cellmapping$cs_type, levels = Celltypes)
ggplot(SWOT_Cellmapping) +
  geom_point(aes(x=cs_x, y=cs_y, color=cs_type), size = 1.3) + 
  scale_color_manual(values = color_Celltype) + 
  theme_bw() +
  labs(x="Cell_X", y="Cell_Y", color='Cell type') +
  ggtitle("Cell-type spatial distribution of single-cell spatial map") +
  theme(plot.title = element_text(hjust=0.5)) +
  #geom_point(aes(x=spot_x, y=spot_y), shape = 21, size = 7.5) +
  guides(color = guide_legend(override.aes = list(shape=15, size = 5))) +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.background = element_blank())

```


### Gene expression levels of single-cell spatial maps
```{r fig.width=10, fig.height=8, message=FALSE}
# Find too 100 markers for each cell type
SC_exp <- read.csv(paste0(data_path, 'PDAC_sc_exp.csv'), header = T, row.names = 1, check.names = F)
SC_meta <- read.csv(paste0(data_path, 'PDAC_sc_meta.csv'), header = T, row.names = 1, check.names = F)
# We used "FindMarkerGenes" funtion in "Visualization_functions.R" to idenfity markers by "Seurat::FindAllMarkers" function.
# markers_top100 <- FindMarkerGenes(SC_exp, SC_meta, n_tops=100, save = FALSE)
# write.csv(markers_top100, paste0(data_path, 'markers_top100.csv'))
markers_top100 <- read.csv(paste0(data_path, 'PDAC_markers_top100.csv'), header = T, row.names = 1, check.names = F) 

# Plot expression levels of top 100 cell-type-specific genes on the inferred single-cell spatial map of PDAC dataset.
SWOT_Cellmapping_xy <- SWOT_Cellmapping[,c('cs_x', 'cs_y')]
library(dplyr)
SWOT_Cellmapping_xy <- rename(SWOT_Cellmapping_xy, c('X' = 'cs_x', 'Y' = 'cs_y'))
Plot_Topmarkergenes(SWOT_Cellmapping_exp, SWOT_Cellmapping_xy, interest_cts = c('Cancer clone A'), 
                    markers_df = markers_top100, n_tops = 100, Size = 1.5, colors = NULL)

```



